src/test/no_pmix_launcher_client.c

```diff
@@ -46,12 +46,101 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <semaphore.h>
-#include <gurt/common.h>
 #include <cart/api.h>
 
+#include "crt_internal.h"
 #include "no_pmix_launcher_common.h"
 #include "tests_common.h"
 
+struct wfr_status {
+	sem_t	sem;
+	int	rc;
+};
+
+static void
+rpc_handle_reply(const struct crt_cb_info *info)
+{
+        sem_t   *sem;
+
+        D_ASSERTF(info->cci_rc == 0, "rpc response failed. rc: %d\n",
+                info->cci_rc);
+
+        sem = (sem_t *)info->cci_arg;
+        sem_post(sem);
+}
+
+static void
+wfr_handle_reply(const struct crt_cb_info *info)
+{
+	struct wfr_status	*wfrs;
+
+	wfrs = (struct wfr_status *)info->cci_arg;
+
+	wfrs->rc = info->cci_rc;
+
+        sem_post(&wfrs->sem);
+}
+
+static int wait_for_ranks(crt_context_t ctx, crt_group_t *grp, d_rank_list_t *rank_list, int tag, double timeout)
+{
+	struct wfr_status	ws;
+	d_rank_t		rank;
+	crt_rpc_t		*rpc = NULL;
+	crt_endpoint_t          server_ep;
+	int                     i = 0;
+	int                     rc = 0;
+
+        rc = sem_init(&ws.sem, 0, 0);
+        D_ASSERTF( rc == 0, "sem_init() failed; rc=%d\n", rc);
+
+        for (i = 0; i < rank_list->rl_nr; i++) {
+
+                rank = rank_list->rl_ranks[i];
+
+                printf("Sending ping to %d:%d\n", rank, tag);
+
+                server_ep.ep_rank = rank;
+                server_ep.ep_tag = tag;
+                server_ep.ep_grp = grp;
+
+                rc = crt_req_create(ctx, &server_ep, CRT_OPC_CTL_GET_PID, &rpc);
+                D_ASSERTF( rc == 0, "crt_req_create() failed; rc=%d\n", rc);
+
+                rc = crt_req_set_timeout(rpc, 5);
+                D_ASSERTF( rc == 0, "crt_req_set_timeout() failed; rc=%d\n", rc);
+
+              	rc = crt_req_send(rpc, wfr_handle_reply, &ws);
+		printf("SCHAN15 - rpc send rc = %d\n", rc);
+               	tc_sem_timedwait(&ws.sem, 10, __LINE__);
+
+	        struct timespec                  t1, t2;
+        	double                           time_s;
+        	rc = d_gettime(&t1);
+        	assert(rc == 0);
+
+		while(ws.rc != 0 && time_s < timeout){
+			printf("SCHAN15 - Retrying...\n");
+			rc = crt_req_create(ctx, &server_ep, CRT_OPC_CTL_GET_PID, &rpc);
+			rc = crt_req_set_timeout(rpc, 5);
+			rc = crt_req_send(rpc, wfr_handle_reply, &ws);
+			tc_sem_timedwait(&ws.sem, 10, __LINE__);
+		        rc = d_gettime(&t2);
+		        assert(rc == 0);
+        		time_s = d_time2s(d_timediff(t1, t2));
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs